src/cpp/transport/discovery/DiscoveryServerWindows.cpp

```diff
@@ -0,0 +1,154 @@
+
+#include <ws2ipdef.h>
+#include <MSWSock.h>
+#include <functional>
+
+#define RECEIVE_TIMEOUT 100
+#define DISCOVERY_IP "239.255.0.2"
+
+namespace eprosima {
+namespace uxr {
+
+DiscoveryServerWindows::DiscoveryServerWindows(const Processor& processor, uint16_t port, uint16_t discovery_port)
+    : DiscoveryServer (processor, port),
+      poll_fd_{},
+      buffer_{0},
+      discovery_port_(discovery_port)
+{
+}
+
+bool DiscoveryServerWindows::init()
+{
+    bool rv = false;
+
+    /* Socket initialization. */
+    poll_fd_.fd = socket(PF_INET, SOCK_DGRAM, IPPROTO_UDP);
+
+    /* Local IP and Port setup. */
+    struct sockaddr_in address;
+    address.sin_family = AF_INET;
+    address.sin_port = htons(discovery_port_);
+    address.sin_addr.s_addr = INADDR_ANY;
+    memset(address.sin_zero, '\0', sizeof(address.sin_zero));
+    if (-1 != bind(poll_fd_.fd, (struct sockaddr*)&address, sizeof(address)))
+    {
+        /* Poll setup. */
+        poll_fd_.events = POLLIN;
+
+        /* Set up multicast IP. */
+        struct ip_mreq mreq;
+        mreq.imr_multiaddr.s_addr = inet_addr(DISCOVERY_IP);
+        mreq.imr_interface.s_addr = INADDR_ANY;
+        if (-1 != setsockopt(poll_fd_.fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, (char*)&mreq, sizeof(mreq)))
+        {
+            /* Get local address. */
+            SOCKET fd = socket(PF_INET, SOCK_DGRAM, 0);
+            struct sockaddr_in temp_addr;
+            temp_addr.sin_family = AF_INET;
+            temp_addr.sin_port = htons(80);
+            temp_addr.sin_addr.s_addr = inet_addr("1.2.3.4");
+            int connected = connect(fd, (struct sockaddr *)&temp_addr, sizeof(temp_addr));
+            if (0 == connected)
+            {
+                struct sockaddr local_addr;
+                int local_addr_len = sizeof(local_addr);
+                if (-1 != getsockname(fd, &local_addr, &local_addr_len))
+                {
+                    transport_address_.medium_locator().address({uint8_t(local_addr.sa_data[2]),
+                                                                 uint8_t(local_addr.sa_data[3]),
+                                                                 uint8_t(local_addr.sa_data[4]),
+                                                                 uint8_t(local_addr.sa_data[5])});
+                    rv = true;
+                }
+                closesocket(fd);
+            }
+        }
+    }
+
+    return rv;
+}
+
+bool DiscoveryServerWindows::close()
+{
+    return (0 == closesocket(poll_fd_.fd));
+}
+
+bool DiscoveryServerWindows::recv_message(InputPacket& input_packet, int timeout)
+{
+    bool rv = true;
+    struct sockaddr client_addr;
+    int client_addr_len = sizeof(client_addr);
+
+    int poll_rv = WSAPoll(&poll_fd_, 1, timeout);
+    if (0 < poll_rv)
+    {
+        int bytes_received = recvfrom(poll_fd_.fd, (char*)buffer_, sizeof(buffer_), 0, &client_addr, &client_addr_len);
+        if (SOCKET_ERROR != bytes_received)
```

Comment:
if `SOCKET_ERROR == bytes_received` the `rv` will be *true* when it must be *false*. Maybe initialize the `rv` variable to *false* and set it to *true* when `SOCKET_ERROR != bytes_received`?

Reply:
You are right! This approach is followed in UDPServer but I forget to change in DiscoveryServer. Thanks!